[Reader] Your Tags Feed

RenanLukas opened the pull request:
Feature branch of Tags IA project (ref: pcdRpT-6hp-p2)

RenanLukas committed:
Remove blog avatar, fix excerpt max lines

RenanLukas committed:
Merge branch 'issue/20587-implement-horizontal-posts-list-ui-component' into issue/20268-implement-horizontal-post-list-ui-shimmer

RenanLukas committed:
Remove blog avatar from loading

RenanLukas committed:
Fix spacing between number of likes and like button in HorizontalPostListItem

RenanLukas committed:
Merge branch 'issue/20587-implement-horizontal-posts-list-ui-component' into issue/20268-implement-horizontal-post-list-ui-shimmer

RenanLukas committed:
Fix detekt, remove extra padding

RenanLukas committed:
Create HorizontalPostList file

RenanLukas committed:
Fix like icon start padding

RenanLukas committed:
Merge branch 'issue/20587-implement-horizontal-posts-list-ui-component' into issue/20268-implement-horizontal-post-list-ui-shimmer

RenanLukas committed:
Merge branch 'issue/20268-implement-horizontal-post-list-ui-shimmer' into issue/20623-implement-horizontal-posts-list-ui-component

Commit:
Create ReaderPostRepository

RenanLukas committed:
Refactor reader tags feed components package

Commit:
Replace StringUtils by kotlin extensions

RenanLukas committed:
Merge pull request #20614 from wordpress-mobile/issue/20587-implement-horizontal-posts-list-ui-component

[Reader] Implement "Tags" feed horizontal posts list item component

RenanLukas committed:
Move shimmer color to generic component

RenanLukas committed:
Merge branch 'issue/20268-implement-horizontal-post-list-ui-shimmer' into issue/20623-implement-horizontal-posts-list-ui-component

RenanLukas committed:
WIP ReaderTagsFeed UI

RenanLukas committed:
WIP Reader tags feed UI states

RenanLukas committed:
Implement ReaderTagsFeed preview

RenanLukas committed:
Change shimmer implementation

RenanLukas committed:
Fix detekt

RenanLukas committed:
Merge branch 'issue/20268-implement-horizontal-post-list-ui-shimmer' into issue/20623-implement-horizontal-posts-list-ui-component

Commit:
Create function to fetch newer posts for tag

Commit:
Create ViewModel methods to fetch posts for Tags Feed

Commit:
Create temporary UI for testing tag fetching

RenanLukas committed:
Add onTagClicked parameter to TagsFeedItem

RenanLukas committed:
Fix horizontal list item spacing

RenanLukas committed:
Implement ReaderTagsFeed loading state

RenanLukas committed:
Implement loading posts and tags loaded state for ReaderTagsFeed

RenanLukas committed:
Remove shimmer effect

RenanLukas committed:
Merge branch 'issue/20268-implement-horizontal-post-list-ui-shimmer' into issue/20623-implement-horizontal-posts-list-ui-component

RenanLukas committed:
Replace ShimmerBox with Box in ReaderTagsFeed

RenanLukas committed:
Fix detekt

RenanLukas committed:
Merge branch 'issue/20268-implement-horizontal-post-list-ui-shimmer' into issue/20623-implement-horizontal-posts-list-ui-component

RenanLukas committed:
Merge pull request #20634 from wordpress-mobile/issue/20268-implement-horizontal-post-list-ui-shimmer

[Reader] Implement "Tags" feed horizontal posts list item component loading

RenanLukas committed:
Fix reader tags feed loading color for dark theme

RenanLukas committed:
Fix reader tag button loading background color

RenanLukas committed:
Refactor ReaderTagsFeed UI state

RenanLukas committed:
WIP ReaderTagsFeed error state

RenanLukas committed:
WIP Update ReaderTagsFeed error state height

Commit:
Fetch content for the actual followed tags

Commit:
Get the Tags Feed tag from Fragment's newInstance

Commit:
Add some doc comments in the Tags Feed ViewModel

RenanLukas committed:
Implement error state for reader tags feed

Commit:
Create a more specific exception for post fetching

Commit:
Create ReaderPostLocalSource for extracting local DB logic

RenanLukas committed:
Implement more from tag button, fix clip to padding

RenanLukas committed:
Implement reader tags feed see more label

RenanLukas committed:
Fix ripple effect on more from tag button

RenanLukas committed:
Refactor ReaderTagsFeed states

RenanLukas committed:
Add click to post title, title excerpt and site name

Commit:
Add unit test for ReaderPostLogicFactory

RenanLukas committed:
Implement ReaderTagsFeed empty state

Commit:
Use ReaderPostTableWrapper in ReaderPostLocalSource

Commit:
Use AppPrefsWrapper in ReaderPostLocalSource

Commit:
Add unit tests for ReaderPostLocalSource

RenanLukas committed:
Update error state messages for ReaderTagsFeed

RenanLukas committed:
Refactor tags feed item to use a specific object instead of a pair

RenanLukas committed:
Merge branch 'issue/20623-implement-horizontal-posts-list-ui-component' into issue/20584-reader-tags-feed-implement-empty-state

RenanLukas committed:
Apply PR suggestion: remove useless Row element

Commit:
Add retry to temporary UI to remove suppress annotation

RenanLukas committed:
Merge pull request #20686 from wordpress-mobile/issue/20623-implement-horizontal-posts-list-ui-component

[Reader] Implement tags feed UI

RenanLukas committed:
Merge pull request #20695 from wordpress-mobile/issue/20584-reader-tags-feed-implement-empty-state

[Reader] Implement "Tags" feed empty state UI

RenanLukas committed:
Merge branch 'trunk' into feature/tags-ia

Commit:
Fix comment typo

Commit:
Add return type to public function

Commit:
Check for returned instance type instead of null in test

Commit:
Add unit tests for ReaderTagsFeedViewModel

Commit:
Fix linebreak format

RenanLukas committed:
WIP replace test UI with real UI in tags feed screen

RenanLukas committed:
WIP Start using UI state from ReaderTagsFeed in ReaderTagsFeedViewModel

RenanLukas committed:
Implement post dateline in tags feed

Commit:
Merge pull request #20684 from wordpress-mobile/issue/20588-tags-feed-fetching

[Tags IA] Fetch posts for multiple tags

RenanLukas committed:
Fix: Add max lines value to post excerpt that has no image

RenanLukas committed:
Change ReaderPostRepository fetch default limit to 10 instead of 7

Commit:
[WIP] Extract SubFilterViewModel init logic to ReaderFragment

Commit:
Change SubFilterViewModelOwner to ...Provider

The idea is that the Provider is the component responsible for fetching the
correct SubFilterViewModel, but it doesn't necessarily own the ViewModel.

This was done so we can use the current feed as the Owner of the SubFilter
ViewModel without it actually declaring it, so the outermost layer of the Reader
(ReaderFragment) is still the source of truth for getting the SubFilterViewModel
but the actual lifecycle of the ViewModel is tied to the current feed fragment.

RenanLukas committed:
Finish integrating ReaderTagsFeed with ReaderTagsFeedViewModel

RenanLukas committed:
Extract loaded TagFeedItem mapping to ReaderTagsFeedUiStateMapper

RenanLukas committed:
Extract loading and error UI states to UI state mapper

RenanLukas committed:
Fix ReaderTagsFeedViewModelTest tests

RenanLukas committed:
Implement ReaderTagsFeedUiStateMapper tests

RenanLukas committed:
Fix detekt

RenanLukas committed:
Implement show tag posts list when tag button is tapped in Tags feed

RenanLukas committed:
Implement show tag posts list when tag button is tapped in Tags feed

Commit:
Merge pull request #20729 from wordpress-mobile/issue/20624-connect-fetch-logic-with-compose-ui

[Reader] Connect fetch tags logic with compose UI

Commit:
Fix lint issue

Commit:
Merge branch 'feature/tags-ia' of github.com:wordpress-mobile/WordPress-Android into issue/subfilter-viewmodel

Commit:
Stop starting the Tags Feed VM multiple times

RenanLukas committed:
Fix number of likes and comments when the count is 0

RenanLukas committed:
Fix bottom navigation view overlapping with compose view

Commit:
Integrate SubFilters with Tag Feed

RenanLukas committed:
WIP open post details action on tags feed

RenanLukas committed:
Implement open post details action in tags feed

Commit:
Fix broken unit test compilation

RenanLukas committed:
Implement open blog details action in tags feed

RenanLukas committed:
Fix detekt

RenanLukas committed:
Fix ReaderTagsFeedViewModelTest unit tests

RenanLukas committed:
Fix ReaderTagsFeedUiStateMapperTest unit tests

Commit:
Make isStarted private again

Commit:
Fix tags feed item site text width

RenanLukas committed:
Update ReaderTagsFeedViewModelTest tests

Commit:
Add unit tests to SubFilterViewModelProvider

RenanLukas committed:
WIP tags feed like action implementation

Commit:
Improve and add unit tests for ReaderTagsFeedViewModel

Commit:
Remove unused import

RenanLukas committed:
WIP update like button UI immediately after click

RenanLukas committed:
Disable like button after is tapped

RenanLukas committed:
Simplify code and add comments

RenanLukas committed:
WIP implement like action in reader tags feed: call like endpoint

RenanLukas committed:
Merge pull request #20747 from wordpress-mobile/issue/20669-implement-more-from-tag-action

[Tags IA] Integrate subfilters and post list in Tags Feed

RenanLukas committed:
Merge branch 'feature/tags-ia' into issue/20670-implement-tags-feed-card-actions

RenanLukas committed:
Merge pull request #20757 from wordpress-mobile/issue/20670-implement-tags-feed-card-actions

[Reader] Implement tags feed card actions (part 1)

Commit:
Implement Lazy Loading in Tags Feed

RenanLukas committed:
Merge branch 'feature/tags-ia' into issue/20670-implement-tags-feed-actions-part-2

RenanLukas committed:
Extract like post remote to separate method

RenanLukas committed:
Implement like button error messages

RenanLukas committed:
WIP update ReaderTagsFeedViewModelTest like button tests

RenanLukas committed:
finish updating ReaderTagsFeedViewModelTest like button tests

RenanLukas committed:
Fix detekt

RenanLukas committed:
Apply PR suggestion: break updatePostItemUI into smaller pieces

RenanLukas committed:
Merge pull request #20779 from wordpress-mobile/issue/20670-implement-tags-feed-actions-part-2

[Reader] Implement tags feed card actions (part 2)

RenanLukas committed:
WIP implement tags feed overflow menu

Commit:
Fix existing unit tests

Commit:
Move post fetching for tags to IO thread

Commit:
Merge branch 'feature/tags-ia' of github.com:wordpress-mobile/WordPress-Android into issue/20708-tags-feed-lazy-load

Commit:
Implment pull-to-refresh in tags feed

Commit:
Show 5 post skeletons in loading state

Commit:
Fix existing unit tests

Commit:
Add unit tests in ViewModel for refresh

Commit:
Show the feature image instead of blog avatar

RenanLukas committed:
WIP implement more menu using existing Reader logic

RenanLukas committed:
Fix more menu button using AndroidView in ReaderTagsFeedPostListItem

Commit:
Improve item update (review comment)

RenanLukas committed:
Implement bookmarked post info dialog

RenanLukas committed:
Implement subscribe action snackbar

RenanLukas committed:
Update unit tests

RenanLukas committed:
Fix detekt

Commit:
Merge pull request #20790 from wordpress-mobile/issue/20708-tags-feed-lazy-load

[Tags Feed] Lazy load tag rows

Commit:
Merge pull request #20791 from wordpress-mobile/issue/20707-tags-feed-ptr

[Tags Feed] Implement pull-to-refresh

RenanLukas committed:
Merge branch 'feature/tags-ia' into issue/20670-implement-tags-feed-actions-part-3

RenanLukas committed:
Fix tags dropdown menu item label

Commit:
Update Loaded item UI

Commit:
Update height of the "more from" button

RenanLukas committed:
Remove unused navigation events from ReaderTagsFeedFragment

Commit:
Update post loading skeleton

RenanLukas committed:
Implement action for more from tag button

Commit:
Improve excerpt rendering and item height

RenanLukas committed:
Implement tags feed empty state button action

RenanLukas committed:
Fix detekt

RenanLukas committed:
Apply PR suggestions

RenanLukas committed:
Remove reblog todo

Commit:
Improve number of title + excerpt lines

RenanLukas committed:
Workaround to update tags feed after changing subscribed status from tag posts list

RenanLukas committed:
Merge pull request #20801 from wordpress-mobile/issue/20670-implement-tags-feed-actions-part-3

[Reader] Implement tags feed card actions (part 3)

dangermattic commented:
<!--
  0 Errors
  5 Warnings: <code>strings.xml</code> files...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          5 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>strings.xml</code> files should only be updated on release branches, when the translations are downloaded by our automation.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>FilterItemType</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>MoreMenuUiState</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-Android/milestone/285">25.0</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

Commit:
Add blog name fallback in case of blank names

Commit:
Change tags feed item content total lines to 3

At most 2 for title, an the rest for excerpt, vertically aligning the content
block between the site tile row and bottom metadata/buttons

Commit:
Update loading skeleton to match new layout

Commit:
Merge branch 'feature/tags-ia' of github.com:wordpress-mobile/WordPress-Android into issue/20788-tags-ia-ui-improvements

Commit:
Bump thin line in skeleton to 10dp

RenanLukas committed:
Implement reader_dropdown_menu_item_tapped event for tags item

RenanLukas committed:
Implement reader_tags_feed_shown analytics event

RenanLukas committed:
Add type parameter to reader_filter_sheet_item_selected event

RenanLukas committed:
Implement reader_tag_header_tapped analytics event

RenanLukas committed:
Update SubFilterViewModel unit tests

RenanLukas committed:
Add type parameter to reader_filter_sheet_cleared event

RenanLukas committed:
Track reader_post_card_tapped event when reader tags feed post is clicked

RenanLukas committed:
Fix tags feed more menu actions source for analytics events

RenanLukas committed:
Track reader_tags_feed_more_from_tag_tapped event when reader tags feed more from tag button is tapped

RenanLukas committed:
Change analytics event name from READER_TAG_HEADER_TAPPED to READER_TAGS_FEED_HEADER_TAPPED

RenanLukas committed:
Change reader filter sheet  events type parameter values

Commit:
Fix unit tests

Commit:
Add proper semantics to loaded item

Commit:
Fix semantics for loading/empty/error states

RenanLukas committed:
Merge pull request #20817 from wordpress-mobile/issue/20585-tags-ia-analytics-events

[Reader] Implement analytics events for Tags IA

Commit:
Merge pull request #20802 from wordpress-mobile/issue/20788-tags-ia-ui-improvements

[Tags Feed] Implement Post Item UI improvements

Commit:
Merge pull request #20825 from wordpress-mobile/issue/20671-reader-tags-feed-a11y

[Tags Feed] Improve accessibility

RenanLukas committed:
Add reader announcement card feature config

Commit:
Implement onRetry function to fetch tag posts

Commit:
Change Error layout height to match post item height

RenanLukas committed:
Add show Reader card announcement methods to AppPrefs

Commit:
Fix and add unit tests for retry

Commit:
Refactor no connection error message string key

Commit:
Implement NoConnection state

Commit:
Improve snackbar code for better positioning

RenanLukas committed:
WIP reader announcement card UI

RenanLukas committed:
Fix reader announcement card text style

RenanLukas committed:
WIP Reader announcement card UI integration

Commit:
Add and update unit tests for no connection scenarios

RenanLukas committed:
Fix scrolling of reader announcement card

RenanLukas committed:
Add real data to reader announcement card

RenanLukas committed:
Implement reader announcement card done button

RenanLukas committed:
Implement done button analytics event

Commit:
Add mechanism to initialize observer only once

Also making sure it works for Fragment recreation by the system (such as screen
rotation) and ViewModel destruction due to owner being destroyed.

Commit:
Fix item selected tracked when first opening Subscriptions/Tags feeds

RenanLukas committed:
Implement unit tests for reader announcement card

RenanLukas committed:
Fix dark theme colors

RenanLukas committed:
Merge branch 'feature/tags-ia' into issue/20621-tags-feed-onboarding

RenanLukas committed:
Fix detekt

Commit:
Revert "Add mechanism to initialize observer only once"

This reverts commit 4855ec7153e7eca9299a0a991c84bef2203c9dc5.

Commit:
Initialize subfilter observers as fields to keep unique reference

This makes sure that the Lifecycle and ViewModel Owners handle everything
correctly when observing LiveData, which avoids duplicate observers and allows
for correct auto removal/re-observation due to lifecycle and owner state change.

Commit:
Update unit tests

Commit:
Merge pull request #20839 from wordpress-mobile/issue/20823-reader-tags-feed-retry

[Tags Feed] Implement retry action for error state

Commit:
Remove unused resource

Commit:
Fix BuildConfig constant being true

Commit:
Replace LazyColumn with Column for announcement items

Commit:
Add the Tags feed announcement item if it's enabled only

Commit:
Merge pull request #20842 from wordpress-mobile/issue/20815-reader-tags-feed-connection-error

[Tags Feed] Implement no connection state

Commit:
Remove unnecessary coroutine in tests

Commit:
Merge branch 'feature/tags-ia' of github.com:wordpress-mobile/WordPress-Android into issue/20814-reader-subfilter-tracking-bug

Commit:
Fix unit tests

Commit:
Merge pull request #20844 from wordpress-mobile/issue/20814-reader-subfilter-tracking-bug

[Tags Feed] Fix duplicate tracking in Reader subfilter

Commit:
Update Loaded state with tags in a diff-style

Keep existing tags content untouched and only add new tags in initial state for
new tags, so the list updates in place.

Also add placement animation to the list items for a better UX.

RenanLukas committed:
Apply PR suggestion: rename loadAnnouncementCard to updateAnnouncementCard

Commit:
Add mapper unit tests

Commit:
Update ReaderTagsFeedViewModel unit tests

RenanLukas committed:
Merge pull request #20846 from wordpress-mobile/issue/20621-tags-feed-onboarding

[Reader] Implement announcement card

RenanLukas committed:
Make reader announcement card scrollable as part of the post list

RenanLukas committed:
Hide reader announcement card on Discover when empty state is shown

RenanLukas committed:
Show reader announcement card only on Discover feed

RenanLukas committed:
Hide reader announcement card when any feed but Discover is selected

RenanLukas committed:
Update ReaderViewModel unit tests

develric committed:
Merge pull request #20861 from wordpress-mobile/issue/20621-tags-feed-onboarding-improvements

[Reader] Implement announcement card improvements

Commit:
Merge pull request #20860 from wordpress-mobile/issue/20811-update-only-tags-list

[Tags Feed] Update feed tags as Diff, keeping existing tag content

Commit:
Use inverted logic since it should only hide Blogs for Tags Feed

Commit:
Merge pull request #20867 from wordpress-mobile/fix/tags-ia-filter-in-filterable-feeds

[Reader] Show filters in A8C and P2 feeds

Commit:
Create ReaderAnnouncementRepository

Commit:
Create ReaderAnnouncementCardView wrapper for Composable

Commit:
Add Announcement card logic to ReaderPostAdapter

Commit:
Add Announcement card to Discover Feed

Commit:
Update and add new unit tests to Discover ViewModel

Commit:
Remove Announcement card from ReaderFragment

Commit:
Update Announcement unit tests

Commit:
Remove the announcement compose view from reader_fragment_layout.xml

Commit:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into feature/tags-ia

Commit:
Improve Item View Type logic

Commit:
Add Reader Announcement on Tags Feed

Commit:
Update mapper tests

Commit:
Update unit tests for ReaderTagsFeedViewModel

Commit:
Fix detekt issue

Commit:
Remove shouldShow from ReaderAnnouncementCard

Commit:
Hide Announcement in Subscriptions feed when filtered

Commit:
Turn on Reader Tags Feed FF by default

Commit:
Turn on Reader Announcement FF by default

Commit:
Rename Announcement repo to ReaderAnnouncementHelper

Commit:
Merge branch 'issue/20621-announcement-card-in-feeds' of github.com:wordpress-mobile/WordPress-Android into issue/20621-announcement-card-in-tags-feed

Commit:
Merge pull request #20872 from wordpress-mobile/issue/20621-announcement-card-in-feeds

[Reader] Move announcement card inside the feeds

RenanLukas committed:
Merge pull request #20877 from wordpress-mobile/issue/20621-announcement-card-in-tags-feed

[Reader] Also add announcement card into the new Tags Feed

RenanLukas committed:
Merge pull request #20879 from wordpress-mobile/issue/reader-tags-announcement-enable-ff

[Tags Feed] Enable FFs by default for Tags Feed and Announcement

RenanLukas committed:
Merge branch 'trunk' into feature/tags-ia

thomashorta reviewed (approved):
Since all commits here were reviewed in other PRs I just took a quick overall look and tested a few scenarios. LGTM!

One minor nitpick: should the PR name be more descriptive, like "[Reader] Your Tags Feed"?
